staging/src/k8s.io/apiserver/pkg/cel/openapi/resolver/refs.go

```diff
@@ -19,19 +19,41 @@ package resolver
 import (
 	"fmt"
 
+	"k8s.io/apimachinery/pkg/util/sets"
 	"k8s.io/kube-openapi/pkg/validation/spec"
 )
 
-// populateRefs recursively replaces Refs in the schema with the referred one.
+// PopulateRefs recursively replaces Refs in the schema with the referred one.
 // schemaOf is the callback to find the corresponding schema by the ref.
 // This function will not mutate the original schema. If the schema needs to be
 // mutated, a copy will be returned, otherwise it returns the original schema.
-func populateRefs(schemaOf func(ref string) (*spec.Schema, bool), schema *spec.Schema) (*spec.Schema, error) {
+func PopulateRefs(schemaOf func(ref string) (*spec.Schema, bool), rootRef string) (*spec.Schema, error) {
+	visitedRefs := sets.New[string]()
+	rootSchema, ok := schemaOf(rootRef)
+	visitedRefs.Insert(rootRef)
+	if !ok {
+		return nil, fmt.Errorf("internal error: cannot resolve Ref for root schema %q: %w", rootRef, ErrSchemaNotFound)
+	}
+	return populateRefs(schemaOf, visitedRefs, rootSchema)
+}
+
+func populateRefs(schemaOf func(ref string) (*spec.Schema, bool), visited sets.Set[string], schema *spec.Schema) (*spec.Schema, error) {
 	result := *schema
 	changed := false
 
 	ref, isRef := refOf(schema)
 	if isRef {
+		if visited.Has(ref) {
+			return &spec.Schema{
+				// for circular ref, return an empty object as placeholder
+				SchemaProps: spec.SchemaProps{Type: []string{"object"}},
+			}, nil
+		}
```

Comment:
I intend to use this API for validating resources from their schemas. This placeholder means that for recursive schemas we don't validate anything past the first recursion.

Would it make sense to keep track of all the circular reference usages and insert them as part of a post-process? AFAICT all usages could share the same instance of `properties`, `items`, `additionalProperties`, etc.

If that is too hacky, and it seems kinda hacky, at least a way to find out what the original reference was to, so that I could write my own code to resolve as I traverse the object tree would work.

Reply:
Could you give me an example of a recursive schema other than JSONSchemaProps? Let's discuss the solution based on use cases.

> Would it make sense to keep track of all the circular reference usages and insert the actual schema as part of a post-process step? AFAICT all usages could share the same instance of properties, items, additionalProperties, etc.

This would require a lazy resolution to the schema fields, which we do not yet have.

> at least a way to find out what the original reference was to, so that I could write my own code to resolve as I traverse the object tree would work.

I think preserving the Ref would do. However, that would mean invading the implementation details of the schema resolver. Again, please give me an example of a type that requires such handling.

Reply:
Actually, no, sorry. It is not allow to have both Ref and a real schema.

Reply:
> Could you give me an example of a recursive schema other than JSONSchemaProps?

I only refer to JSONSchemaProps. I'm trying to validate CRDs from their schemas.

> Again, please give me an example of a type that requires such handling.

JSONSchemaProps

> However, that would mean invading the implementation details of the schema resolver.

This is not urgent, so I can look into adding later. Thanks